reciprocalspaceship/io/crystfel.py

```diff
@@ -0,0 +1,190 @@
+        return s.startswith("Image filename")
+
+    def contains_event(s):
+        return s.startswith("Event")
+
+    def contains_serial_number(s):
+        return s.startswith("Image serial number")
+
+    def starts_chunk_peaks(s):
+        return s.startswith("  fs/px   ss/px (1/d)/nm^-1   Intensity  Panel")
+
+    def ends_chunk_peaks(s):
+        return s.startswith("End of peak list")
+
+    def starts_crystal_peaks(s):
+        return s.startswith(
+            "   h    k    l          I   sigma(I)       peak background  fs/px  ss/px panel"
+        )
+
+    def is_photon_energy(s):
+        return s.startswith('photon_energy_eV')
+
+    def is_astar(s):
+        return s.startswith('astar')
+
+    def is_bstar(s):
+        return s.startswith('bstar')
+
+    def is_cstar(s):
+        return s.startswith('cstar')
+
+    def ends_crystal_peaks(s):
+        return s.startswith("End of reflections")
+
+    def eV2Angstrom(e_eV):
+        return 12398. / e_eV
+
+    with open(filename, "r") as stream:
+        is_chunk = False
+        is_crystal = False
+        current_filename = None
+        current_event = None  # to handle non-event streams
+        current_serial_number = None
+        corrupted_chunk = False
+        crystal_peak_number = 0
+        crystal_idx = 0
+
+        for line in stream:
+            # analyzing what we have
+            if ends_chunk_peaks(line):
+                is_chunk = False
+                chunk_peak_number = 0
+            elif ends_crystal_peaks(line):
+                is_crystal = False
+                crystal_peak_number = 0
+
+            elif is_photon_energy(line):
+                photon_energy = float(line.split()[2])
+            elif is_astar(line):
+                astar = np.array(list(map(
+                    float,
+                    line.split()[2:5]))) / 10.  # crystfel's notation uses nm-1
+            elif is_bstar(line):
+                bstar = np.array(list(map(
+                    float,
+                    line.split()[2:5]))) / 10.  # crystfel's notation uses nm-1
+            elif is_cstar(line):
+                cstar = np.array(list(map(
+                    float,
+                    line.split()[2:5]))) / 10.  # crystfel's notation uses nm-1
+
+                # since it's the last line needed to construct Ewald offset,
+                # we'll pre-compute the matrices here
+                A = np.array([astar, bstar, cstar]).T
+                lambda_inv = 1 / eV2Angstrom(photon_energy)
+                s0 = np.array([0, 0, lambda_inv]).T
+
+            elif is_crystal:
+                # example line:
+                #    h    k    l          I   sigma(I)       peak background  fs/px  ss/px panel
+                #  -63   41    9     -41.31      57.45     195.00     170.86  731.0 1350.4 p0
+                crystal_peak_number += 1
+                h, k, l, I, sigmaI, _, _, _, _, _ = [i for i in line.split()]
+                h, k, l = map(float, [h, k, l])
```

Comment:
Should `h`, `k`, and `l` be cast to `int`? Or is it necessary to have floats for the subsequent matrix multply?

Reply:
You're right, `int` should be enough. `numpy` casts them to float during the calculation.